lib/auth/in_memory/AuthLoader.js

```diff
@@ -0,0 +1,231 @@
+ */
+class AuthLoader {
+    constructor(logApi, checkSas) {
+        this._log = new (logApi || werelogs).Logger('S3');
+        this._authData = { accounts: [] };
+        // null: unknown validity, true/false: valid or invalid
+        this._isValid = null;
+
+        this._joiKeysValidator = joi.array()
+            .items({
+                access: joi.string().required(),
+                secret: joi.string().required(),
+            })
+            .required();
+
+        let sasToken = joi.string();
+        if (checkSas) {
+            sasToken = sasToken.required();
+        }
+        let accountsJoi = joi.array()
+                .items({
+                    name: joi.string().required(),
+                    email: joi.string().email().required(),
+                    arn: joi.string().required(),
+                    canonicalID: joi.string().required(),
+                    shortid: joi.string().regex(/^[0-9]{12}$/).required(),
+                    sasToken,
+                    keys: this._joiKeysValidator,
+                    // backward-compat
+                    users: joi.array(),
+                })
+                .required()
+                .unique('arn')
+                .unique('email')
+                .unique('canonicalID');
+        if (checkSas) {
+            accountsJoi = accountsJoi.unique('sasToken');
+        }
+        this._joiValidator = joi.object({ accounts: accountsJoi });
+    }
+
+    /**
+     * add one or more accounts to the authentication info
+     *
+     * @param {object} authData - authentication data
+     * @param {object[]} authData.accounts - array of account data
+     * @param {string} authData.accounts[].name - account name
+     * @param {string} authData.accounts[].email: email address
+     * @param {string} authData.accounts[].arn: account ARN,
+     *   e.g. 'arn:aws:iam::123456789012:root'
+     * @param {string} authData.accounts[].canonicalID account
+     *   canonical ID
+     * @param {string} authData.accounts[].shortid account ID number,
+     *   e.g. '123456789012'
+     * @param {string} authData.accounts[].sasToken SAS token for
+     *   Azure
+     * @param {object[]} authData.accounts[].keys array of
+     *   access/secret keys
+     * @param {object[]} authData.accounts[].keys[].access access key
+     * @param {object[]} authData.accounts[].keys[].secret secret key
+     * @param {string} [filePath] - optional file path info for
+     *   logging purpose
+     * @return {undefined}
+     */
+    addAccounts(authData, filePath) {
+        const isValid = this._validateData(authData, filePath);
+        if (isValid) {
+            this._authData.accounts =
+                this._authData.accounts.concat(authData.accounts);
+            // defer validity checking when getting data to avoid
+            // logging multiple times the errors (we need to validate
+            // all accounts at once to detect duplicate values)
+            if (this._isValid) {
```

Comment:
The `AuthLoader` object could be used live and updated during the lifetime of the process (e.g. for Pensieve), if the user has requested the data before, it may have been validated, so we have to reset to `null` to force a full revalidation next time the user requests the validated data. But now a `false` validity value is permanent, we may need to revisit this so to recover from errors gracefully instead of having to regenerate a new `AuthLoader` object from scratch.